### ISSUE TITLE
Swift API doesn't work

### ISSUE THREAD

**Issue by akankshajain** [blueboxgroup/ursula#1157]
Swift installation is successful. I had to create swift : service,user,endpoint urls in keystone database manually. Here are the endpoint URLs I have created.
+-------------+-------------------------------------------------+
|   Property  |                      Value                      |
+-------------+-------------------------------------------------+
|   adminurl  | https://swift.example.org:8090/v2/%(tenant_id)s |
|      id     |         675807c708354b90a09f5d0859d20a89        |
| internalurl | https://swift.example.org:8090/v2/%(tenant_id)s |
|  publicurl  | https://swift.example.org:8090/v2/%(tenant_id)s |
|    region   |                    regionOne                    |
|  service_id |         39578ba2e5cf4f0d997a67be6dbe749c        |
+-------------+-------------------------------------------------+

I am trying to run command "swift stat" from controller node , and I am getting:
hostname 'swift.example.org' doesn't match u'openstack.example.org'

**Comment by ryshah** [comment]
@akankshajain If you are using self-signed certificate which is most likely the case, you will need to add insecure = true flag in /etc/swift/proxy-server.conf and restart proxy server "swift-init restart all". After that swift --insecure stat command should work. 

I believe @j2sol had looked in to insecure flag for one of his keystone pull request and can provide more guidance.

Regarding creating keystone endpoint, there is another PR #1143 in progress which will create that as part of deployment

**Comment by j2sol** [comment]
Ursula's swift deployment expects swift end points to be running their own front end, with their own FQDN, which would mean their own SSL cert as well. And yes, if using self signed certs, you either need to have the CA imported into your ssl stack, or use the insecure flag.

**Comment by akankshajain** [comment]
Thanks for the pointer @ryshah and @j2sol . This worked and I was able to resolve the certificate error but now when I hit the swift stat command I get "Account HEAD failed: https://swift.example.org:8090/v2/060a734844194497a23b02d2a24c7edd 401 Unauthorized".

I have created a user named "swift" with password "asdf" and provided it admin role. Here is my identity section in proxy-server.conf

[filter:authtoken]
paste.filter_factory = keystonemiddleware.auth_token:filter_factory
identity_uri = https://openstack.example.org:35358/
auth_uri = https://openstack.example.org:5001/
admin_password = asdf
admin_tenant_name = service
admin_user = swift
cache = swift.cache
include_service_catalog = False
delay_auth_decision = true

Appreciate your help in providing pointers.
Thanks again!

**Comment by ryshah** [comment]
@akankshajain What is the output from swift --debug stat command? Also your stackrc/openrc sets up environment correctly I assume

**Comment by akankshajain** [comment]
@ryshah Here is my stackrc file content
export OS_PASSWORD=ghij
export OS_AUTH_URL=https://openstack.example.org:5001/v2.0
export OS_USERNAME=provider_admin
export OS_TENANT_NAME=admin
export OS_CACERT=/opt/stack/ssl/openstack.crt
export OS_NO_CACHE=True
export OS_VOLUME_API_VERSION=2

Here is my --debug output

DEBUG:requests.packages.urllib3.connectionpool:"POST /v2.0/tokens HTTP/1.1" 200 4168
DEBUG:iso8601.iso8601:Parsed 2015-09-03T04:54:59Z into {'tz_sign': None, 'second_fraction': None, 'hour': u'04', 'daydash': u'03', 'tz_hour': None, 'month': None, 'timezone': u'Z', 'second': u'59', 'tz_minute': None, 'year': u'2015', 'separator': u'T', 'monthdash': u'09', 'day': None, 'minute': u'54'} with default timezone <iso8601.iso8601.Utc object at 0x7f5398346790>
DEBUG:iso8601.iso8601:Got u'2015' for 'year' with default None
DEBUG:iso8601.iso8601:Got u'09' for 'monthdash' with default 1
DEBUG:iso8601.iso8601:Got 9 for 'month' with default 9
DEBUG:iso8601.iso8601:Got u'03' for 'daydash' with default 1
DEBUG:iso8601.iso8601:Got 3 for 'day' with default 3
DEBUG:iso8601.iso8601:Got u'04' for 'hour' with default None
DEBUG:iso8601.iso8601:Got u'54' for 'minute' with default None
DEBUG:iso8601.iso8601:Got u'59' for 'second' with default None
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): swift.example.org
/usr/local/lib/python2.7/dist-packages/requests/packages/urllib3/connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)
DEBUG:requests.packages.urllib3.connectionpool:"HEAD /v2/060a734844194497a23b02d2a24c7edd HTTP/1.1" 401 0
INFO:swiftclient:REQ: curl -i https://swift.example.org:8090/v2/060a734844194497a23b02d2a24c7edd -I -H "X-Auth-Token: cbc819152cd7468d93df73b66937c9bb"
INFO:swiftclient:RESP STATUS: 401 Unauthorized
INFO:swiftclient:RESP HEADERS: [('content-length', '0'), ('connection', 'close'), ('x-trans-id', 'tx38c306f87f1c4439a6687-0055e68123'), ('date', 'Wed, 02 Sep 2015 04:54:59 GMT'), ('content-type', 'text/html; charset=UTF-8'), ('www-authenticate', 'Swift realm="060a734844194497a23b02d2a24c7edd", Keystone uri=\'https://openstack.example.org:5001/v2.0\'')]
DEBUG:keystoneclient.auth.identity.v2:Making authentication request to https://openstack.example.org:5001/v2.0/tokens
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): openstack.example.org
/usr/local/lib/python2.7/dist-packages/requests/packages/urllib3/connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)
DEBUG:requests.packages.urllib3.connectionpool:"POST /v2.0/tokens HTTP/1.1" 200 4168
DEBUG:iso8601.iso8601:Parsed 2015-09-03T04:55:00Z into {'tz_sign': None, 'second_fraction': None, 'hour': u'04', 'daydash': u'03', 'tz_hour': None, 'month': None, 'timezone': u'Z', 'second': u'00', 'tz_minute': None, 'year': u'2015', 'separator': u'T', 'monthdash': u'09', 'day': None, 'minute': u'55'} with default timezone <iso8601.iso8601.Utc object at 0x7f5398346790>
DEBUG:iso8601.iso8601:Got u'2015' for 'year' with default None
DEBUG:iso8601.iso8601:Got u'09' for 'monthdash' with default 1
DEBUG:iso8601.iso8601:Got 9 for 'month' with default 9
DEBUG:iso8601.iso8601:Got u'03' for 'daydash' with default 1
DEBUG:iso8601.iso8601:Got 3 for 'day' with default 3
DEBUG:iso8601.iso8601:Got u'04' for 'hour' with default None
DEBUG:iso8601.iso8601:Got u'55' for 'minute' with default None
DEBUG:iso8601.iso8601:Got u'00' for 'second' with default None
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): swift.example.org
/usr/local/lib/python2.7/dist-packages/requests/packages/urllib3/connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)
DEBUG:requests.packages.urllib3.connectionpool:"HEAD /v2/060a734844194497a23b02d2a24c7edd HTTP/1.1" 401 0
INFO:swiftclient:REQ: curl -i https://swift.example.org:8090/v2/060a734844194497a23b02d2a24c7edd -I -H "X-Auth-Token: 8f533937967d481d873a15ce05f8c1c2"
INFO:swiftclient:RESP STATUS: 401 Unauthorized
INFO:swiftclient:RESP HEADERS: [('content-length', '0'), ('connection', 'close'), ('x-trans-id', 'txd4b655fbc1e346b2b99ce-0055e68124'), ('date', 'Wed, 02 Sep 2015 04:55:00 GMT'), ('content-type', 'text/html; charset=UTF-8'), ('www-authenticate', 'Swift realm="060a734844194497a23b02d2a24c7edd", Keystone uri=\'https://openstack.example.org:5001/v2.0\'')]
ERROR:swiftclient:Account HEAD failed: https://swift.example.org:8090/v2/060a734844194497a23b02d2a24c7edd 401 Unauthorized
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/swiftclient/client.py", line 1302, in _retry
    rv = func(self.url, self.token, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/swiftclient/client.py", line 581, in head_account
    http_response_content=body)
ClientException: Account HEAD failed: https://swift.example.org:8090/v2/060a734844194497a23b02d2a24c7edd 401 Unauthorized
Account HEAD failed: https://swift.example.org:8090/v2/060a734844194497a23b02d2a24c7edd 401 Unauthorized

**Comment by akankshajain** [comment]
I have not changed any configuration files / openrc files after installation.

**Comment by akankshajain** [comment]
I tries changing my stackrc to 

export OS_PASSWORD=asdf
export OS_AUTH_URL=https://openstack.example.org:5001/v2.0
export OS_USERNAME=swift
export OS_TENANT_NAME=service
export OS_NO_CACHE=True
export OS_VOLUME_API_VERSION=2

Got similar error
DEBUG:requests.packages.urllib3.connectionpool:"POST /v2.0/tokens HTTP/1.1" 200 4096
DEBUG:iso8601.iso8601:Parsed 2015-09-03T05:21:32Z into {'tz_sign': None, 'second_fraction': None, 'hour': u'05', 'daydash': u'03', 'tz_hour': None, 'month': None, 'timezone': u'Z', 'second': u'32', 'tz_minute': None, 'year': u'2015', 'separator': u'T', 'monthdash': u'09', 'day': None, 'minute': u'21'} with default timezone <iso8601.iso8601.Utc object at 0x7f3625e1c790>
DEBUG:iso8601.iso8601:Got u'2015' for 'year' with default None
DEBUG:iso8601.iso8601:Got u'09' for 'monthdash' with default 1
DEBUG:iso8601.iso8601:Got 9 for 'month' with default 9
DEBUG:iso8601.iso8601:Got u'03' for 'daydash' with default 1
DEBUG:iso8601.iso8601:Got 3 for 'day' with default 3
DEBUG:iso8601.iso8601:Got u'05' for 'hour' with default None
DEBUG:iso8601.iso8601:Got u'21' for 'minute' with default None
DEBUG:iso8601.iso8601:Got u'32' for 'second' with default None
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): swift.example.org
/usr/local/lib/python2.7/dist-packages/requests/packages/urllib3/connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)
DEBUG:requests.packages.urllib3.connectionpool:"HEAD /v2/a0c4f2375b4c41d899a4767cd1e0f785 HTTP/1.1" 401 0
INFO:swiftclient:REQ: curl -i https://swift.example.org:8090/v2/a0c4f2375b4c41d899a4767cd1e0f785 -I -H "X-Auth-Token: f14ce4cee6bd4bd7ac5862e957b265cf"
INFO:swiftclient:RESP STATUS: 401 Unauthorized
INFO:swiftclient:RESP HEADERS: [('content-length', '0'), ('connection', 'close'), ('x-trans-id', 'tx0a2bb69a25e44406b8fa4-0055e6875c'), ('date', 'Wed, 02 Sep 2015 05:21:32 GMT'), ('content-type', 'text/html; charset=UTF-8'), ('www-authenticate', 'Swift realm="a0c4f2375b4c41d899a4767cd1e0f785", Keystone uri=\'https://openstack.example.org:5001/v2.0\'')]
ERROR:swiftclient:Account HEAD failed: https://swift.example.org:8090/v2/a0c4f2375b4c41d899a4767cd1e0f785 401 Unauthorized
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/swiftclient/client.py", line 1302, in _retry
    rv = func(self.url, self.token, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/swiftclient/client.py", line 581, in head_account
    http_response_content=body)
ClientException: Account HEAD failed: https://swift.example.org:8090/v2/a0c4f2375b4c41d899a4767cd1e0f785 401 Unauthorized
Account HEAD failed: https://swift.example.org:8090/v2/a0c4f2375b4c41d899a4767cd1e0f785 401 Unauthorized

**Comment by ryshah** [comment]
@akankshajain Try adding "cafile= <path to your crt>" parameter in [filter:authtoken] section in /etc/swift/proxy-server.conf and restart the service.

I was able to recreate the issue you are having when my conf didn't have cafile set

**Comment by akankshajain** [comment]
@ryshah I pulled out the latest code from master branch and installed swift topology again. This time swift entities were created automatically in the database.

My [filter:authtoken] section in /etc/swift/proxy-server.conf looks like this now

paste.filter_factory = keystonemiddleware.auth_token:filter_factory
identity_uri = https://openstack.example.org:35358/
auth_uri = https://openstack.example.org:5001/
admin_password = asdf
admin_tenant_name = service
admin_user = swift
cache = swift.cache
include_service_catalog = False
delay_auth_decision = true
cafile = /usr/local/share/ca-certificates/openstack.example.org.crt

But still I am getting 401 unauthorized.

My stackrc file looks like this -
export OS_PASSWORD=ghij
export OS_AUTH_URL=https://openstack.example.org:5001/v2.0
export OS_USERNAME=provider_admin
export OS_TENANT_NAME=admin
export OS_CACERT=/opt/stack/ssl/openstack.crt
export OS_NO_CACHE=True
export OS_VOLUME_API_VERSION=2

**Comment by ryshah** [comment]
Hmmm... Curious if you checked whether /opt/stack/ssl/openstack.crt and  /usr/local/share/ca-certificates/openstack.example.org.crt are same?

**Comment by akankshajain** [comment]
Exactly the same on both controller as well as swift node. But I am using --insecure while running swift commands.

I have not edited ursula/envs/example/swift/group_vars/all.yml
I can see section 

# NOTE: The following private key is *not safe* you should replace
# this with your own private key distribution mechanism. Here for
# demonstration purposes only!

Is this creating issue?

**Comment by akankshajain** [comment]
Hi @ryshah / @j2sol  ,

Any pointers for fixing this issue?

**Comment by ryshah** [comment]
@akankshajain I will try to bootstrap swift env today from master to see if i can recreate the issue and provide you with list of things I had to do to get it work...

**Comment by akankshajain** [comment]
@ryshah : Thanks.
Just lisitng down the steps to help you reproduce the defect
- installed swift topology
- On swift node1:
         Editted /etc/swift/proxy-server.conf to add "insecure = true" as last line in [filter:authtoken]section.
Restart swift proxy service using cmd - swift-init restart all
- Login to controller node
- source /root/stackrc
- executed "swift --debug --insecure stat"

**Comment by ryshah** [comment]
@akankshajain I setup 2 node Swift cluster with one node as controller and other node as Swift node. I started all the services on Swift after deploy was complete. Executing "swift --insecure stat" worked fine for me (without making change to /etc/swift/proxy-server.conf). I added insecure = true to the conf file. And Swift worked fine with that as well.

I got the latest ursula from master and had OS reloaded the nodes before starting deploy. Is that an option for you i.e. OS reload nodes before deploy?

**Comment by akankshajain** [comment]
@ryshah : This is so strange. I again pulled the latest code from master branch today and built a swift setup.

Started all swift services on swiftnode1 using "swift-init all start"
and ran "swift --insecure stat" on controller node. Got 401 unauthorized error again :(

Could you please tell me what do you mean by OS reload? What are the steps for that? Do you think that could be the issue?

Does your stackrc file look like this ? with the same password and user name? I assume you are sourcing the same file before running swift commands?

export OS_PASSWORD=ghij
export OS_AUTH_URL=https://openstack.example.org:5001/v2.0
export OS_USERNAME=provider_admin
export OS_TENANT_NAME=admin
export OS_CACERT=/opt/stack/ssl/openstack.crt
export OS_NO_CACHE=True
export OS_VOLUME_API_VERSION=2

My "swift-init all status" command results in 2 services not running and I do "sudo su" to switch to root user for running any swift-init command I hope that's expected-
No object-reconstructor running
No container-reconciler running

One more point I would like to bring to your notice is I had to change vagrant.yml file of swift node to change "SATA Controller" to "IDE Controller" as my Ubuntu-trusty .box file had this -
 <StorageController name="IDE Controller" type="PIIX4" PortCount="2" useHostIOCache="true" Bootable="true">

Without this I was getting error bootstrapping the swift node. Though I don't think this should be the issue but still.

**Comment by ryshah** [comment]
@akankshajain I was bootstrapping bare metal nodes and not vagrant. I was able to recreate 401 in vagrant. You just need to add "insecure = false" in proxy-server.conf and restart the server

**Comment by akankshajain** [comment]
@ryshah Is there any way I can get things working on vagrant? 

adding "insecure = false" in proxy-server.conf and then restarting swift services did not work on vagrant setup atleast. Is this step required for the bare metal setup?

**Comment by ryshah** [comment]
@akankshajain Sorry I had typo .. It should be "insecure = true" in proxy-server.conf
For bare metal setup it is required unless you commercial certificate or import self-signed in your CA like @j2sol mentioned

**Comment by akankshajain** [comment]
@ryshah : Thanks for clearing out.
So there is no way we can get this working on vagrant?

**Comment by ryshah** [comment]
@akankshajain No it should work in vagrant....I deployed vagrant yesterday saw 401 unauthorized issue.... Added insecure=true flag and that started working... If that still doesn't work for you, can you add contents of /var/log/haproxy.log?

**Comment by j2sol** [comment]
@akankshajain Make sure you have insecure=true BOTH in your stackrc file /and/ in the swift-proxy config file, within the authtoken section.

**Comment by akankshajain** [comment]
@ryshah , @j2sol 

I added insecure=true in proxy conf file
[filter:authtoken]
paste.filter_factory = keystonemiddleware.auth_token:filter_factory
identity_uri = https://openstack.example.org:35358/
auth_uri = https://openstack.example.org:5001/
admin_password = asdf
admin_tenant_name = service
admin_user = swift
cache = swift.cache
include_service_catalog = False
delay_auth_decision = true
cafile = /usr/local/share/ca-certificates/openstack.example.org.crt
insecure = true

I execute "swift --insecure stat". Do I still need to add insecure=true in stackrc?
My /var/log/haproxy.log is blank.

I also added insecure=true in starckrc file and then tried but still same error
Account HEAD failed: https://openstack.example.org:8090/v1/de18f587f1fb4bb0b3f650c31e2d34d9 401 Unauthorized

Here is my stackrc file after adding insecure=true bit it doesn't work  even if its not there.
export OS_PASSWORD=ghij
export OS_AUTH_URL=https://openstack.example.org:5001/v2.0
export OS_USERNAME=provider_admin
export OS_TENANT_NAME=admin
export OS_CACERT=/opt/stack/ssl/openstack.crt
export OS_NO_CACHE=True
export OS_VOLUME_API_VERSION=2
insecure=true

**Comment by j2sol** [comment]
Is this still an issue?

**Comment by akankshajain** [comment]
We can close this issue. Thanks for all the pointers!

Thanks and Regards,
Akanksha
